service/gcs/bridge/bridge.go

```diff
@@ -14,385 +16,563 @@ import (
 	"github.com/Microsoft/opengcs/service/gcs/prot"
 	"github.com/Microsoft/opengcs/service/gcs/transport"
 	"github.com/Microsoft/opengcs/service/libs/commonutils"
-	"github.com/sirupsen/logrus"
 	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
 )
 
-// bridge defines the bridge client in the GCS.
-type bridge struct {
-	// tport is the Transport interface used by the bridge.
-	tport transport.Transport
+// NotSupported represents the default handler logic for an unmatched
+// request type sent from the bridge.
+func NotSupported(w ResponseWriter, r *Request) {
+	w.Error(gcserr.WrapHresult(errors.Errorf("bridge: function not supported, header type: 0x%x", r.Header.Type), gcserr.HrNotImpl))
+}
 
-	// coreint is the Core interface used by the bridge.
-	coreint core.Core
+// NotSupportedHandler creates a default HandlerFunc out of
+// the NotSupported handler logic
+func NotSupportedHandler() Handler {
+	return HandlerFunc(NotSupported)
+}
 
-	// commandConn is the Connection the bridge receives commands (such as
-	// ComputeSystemCreate) over.
-	commandConn transport.Connection
+// Handler responds to a bridge request.
+type Handler interface {
+	ServeMsg(ResponseWriter, *Request)
+}
+
+// HandlerFunc is an adapter to use functions as handlers
+type HandlerFunc func(ResponseWriter, *Request)
+
+// ServeMsg calls f(w, r)
+func (f HandlerFunc) ServeMsg(w ResponseWriter, r *Request) {
+	f(w, r)
+}
+
+// Mux is a protocol multiplexer for request response pairs
+// following the bridge protocol.
+type Mux struct {
+	mu sync.Mutex
+	m  map[prot.MessageIdentifier]Handler
+}
 
-	// writeLock must be held while writing to the transport to ensure that
-	// message data is kept together when writing from multiple go routines
-	// simultaneously.
-	writeLock sync.Mutex
+// NewBridgeMux creates a default bridge multiplexer
+func NewBridgeMux() *Mux {
+	return new(Mux)
 }
 
-// NewBridge produces a new bridge struct using the given Transport and Core
-// interfaces.
-func NewBridge(tport transport.Transport, coreint core.Core) *bridge {
-	return &bridge{
-		tport:   tport,
-		coreint: coreint,
+// Handle registers the handler for the giveen message id
+func (mux *Mux) Handle(id prot.MessageIdentifier, handler Handler) {
+	mux.mu.Lock()
+	defer mux.mu.Unlock()
+
+	if handler == nil {
+		panic("bridge: nil handler")
+	}
+
+	if mux.m == nil {
```

Comment:
I don't feel super strongly, but it does seem better to me to guarantee the map is initialized on construction. That way we don't need any nil checks. The map is pretty much always going to be initialized anyway, and creating an empty map sooner shouldn't really be a performance problem.